java/ql/src/experimental/Security/CWE/CWE-600/UncaughtServletException.ql

```diff
@@ -0,0 +1,85 @@
+/**
+ * @name Uncaught Servlet Exception
+ * @description Uncaught exceptions in a servlet could leave a system in a vulnerable state, possibly resulting in denial-of-service attacks or the exposure of sensitive information disclosed in stack traces.
+ * @kind path-problem
+ * @id java/uncaught-servlet-exception
+ * @tags security
+ *       external/cwe-600
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.frameworks.Servlets
+import semmle.code.xml.WebXML
+import DataFlow::PathGraph
+
+/** The type `java.io.IOException`. */
+class IOException extends RefType {
+  IOException() { this.hasQualifiedName("java.io", "IOException") }
+}
+
+/** Check whether a given exception type is caught. */
+private predicate catchesEx(TryStmt t, RefType exType) {
+  exists(CatchClause cc, LocalVariableDeclExpr v |
+    t.getACatchClause() = cc and
+    cc.getVariable() = v and
+    v.getType().(RefType).getASubtype*() = exType //Detect the case that a subclass exception is thrown but its parent class is declared in the catch clause.
+  )
+}
+
+/** Servlet methods of `javax.servlet.http.Servlet` and subtypes. */
+private predicate isServletMethod(Callable c) {
+  c.getDeclaringType() instanceof ServletClass and
+  c.getNumberOfParameters() = 2 and
+  c.getParameter(1).getType() instanceof ServletResponse and
+  (
+    c.getName() = "doGet" or
+    c.getName() = "doPost" or
+    c.getName() = "doPut" or
+    c.getName() = "doDelete" or
+    c.getName() = "doHead" or
+    c.getName() = "doOptions" or
+    c.getName() = "doTrace" or
+    c.getName() = "service"
+  )
+}
+
+/** Holds if `web.xml` has an error page configured. */
+private predicate hasErrorPage() {
+  exists(WebErrorPage wep | wep.getPageLocation().getValue() != "")
+}
+
+/** Sink of uncaught IO exceptions or runtime exceptions since other exception types must be explicitly caught. */
+class UncaughtServletExceptionSink extends DataFlow::ExprNode {
+  UncaughtServletExceptionSink() {
+    exists(Method m, MethodAccess ma | ma.getMethod() = m |
+      isServletMethod(ma.getEnclosingCallable()) and
+      (
+        m.getAThrownExceptionType().getASupertype*() instanceof IOException or
+        m
+            .getAThrownExceptionType()
+            .getASupertype*()
+            .hasQualifiedName("java.lang", "RuntimeException")
+      ) and
+      ma.getAnArgument() = this.getExpr() and
+      not exists(TryStmt t |
+        t.getBlock() = ma.getEnclosingStmt().getEnclosingStmt*() and
```

Comment:
```suggestion
        t.getBlock() = ma.getEnclosingStmt+() and
```

Reply:
The suggested change is more concise and shall function the same. However, it won't execute and gives an error "Relation argument types (Expr and Stmt) must be compatible for transitive closure". Sorry for my ignorance here, how to fix this grammar issue?

Reply:
ah! Yes you're right, that's because the first one is rising from an expr to a stmt, then the second one is stmt -> stmt, so no need for any change.

Reply:
Thanks for the confirmation.

Reply:
Note that #3687 also added `getAnEnclosingStmt()` as shortcut.